app/maptools/measurementmaptool.cpp

```diff
@@ -0,0 +1,274 @@
+
+void MeasurementMapTool::addPoint()
+{
+  QgsPoint transformedPoint = mapSettings()->screenToCoordinate( mCrosshairPoint );
+  mPoints.push_back( transformedPoint );
+  rebuildGeometry();
+}
+
+void MeasurementMapTool::removePoint()
+{
+  if ( !mPoints.isEmpty() )
+  {
+    mPoints.pop_back();
+    checkCanCloseShape();
+    rebuildGeometry();
+  }
+}
+
+void MeasurementMapTool::updateDistance()
+{
+  if ( mPoints.isEmpty() )
+  {
+    setLengthWithGuideline( 0.0 );
+    return;
+  }
+
+  checkCanCloseShape();
+
+  QgsPoint lastPoint = mPoints.last();
+  QgsPoint transformedCrosshairPoint = mapSettings()->screenToCoordinate( mCrosshairPoint );
+
+  double calculatedLength = mDistanceArea.measureLength( mRecordedGeometry ) + mDistanceArea.measureLine( transformedCrosshairPoint, lastPoint );
+  setLengthWithGuideline( calculatedLength );
+}
+
+void MeasurementMapTool::checkCanCloseShape()
+{
+  if ( mRecordedGeometry.isEmpty() || mPoints.count() < 3 )
+  {
+    setCanCloseShape( false );
+    return;
+  }
+
+  if ( mPoints.count() >= 3 )
+  {
+    QgsPoint firstPoint = mPoints.first();
+    QPointF firstPointScreen = mapSettings()->coordinateToScreen( firstPoint );
+    double distanceToFirstPoint = std::hypot( mCrosshairPoint.x() - firstPointScreen.x(), mCrosshairPoint.y() - firstPointScreen.y() );
+    setCanCloseShape( distanceToFirstPoint <= CLOSE_THRESHOLD );
+  }
+}
+
+void MeasurementMapTool::closeShape()
+{
+  if ( mPoints.count() < 3 )
+    return;
+
+  QList<QgsPointXY> pointList;
+  for ( const QgsPoint &point : mPoints )
+    pointList.append( QgsPointXY( point.x(), point.y() ) );
+
+  QgsGeometry polygonGeometry = QgsGeometry::fromPolygonXY( QList<QList<QgsPointXY>>() << pointList );
+  setRecordedGeometry( polygonGeometry );
+
+  double area = mDistanceArea.measureArea( polygonGeometry );
+  setArea( area );
+
+  double perimeter = mDistanceArea.measurePerimeter( polygonGeometry );
+  setPerimeter( perimeter );
+
+  setCanCloseShape( false );
+  setCloseShapeDone( true );
+}
+
+void MeasurementMapTool::reset()
+{
+  mPoints.clear();
+
+  setPerimeter( 0.0 );
+  setLengthWithGuideline( 0.0 );
+  setCanCloseShape( false );
+  setCloseShapeDone( false );
+
+  rebuildGeometry();
+}
+
+void MeasurementMapTool::rebuildGeometry()
+{
+  QgsGeometry geometry;
+
+  QgsMultiPoint *existingVertices = new QgsMultiPoint();
+  mExistingVertices.set( existingVertices );
+
+  if ( mPoints.count() > 0 )
+  {
+    geometry = QgsGeometry::fromPolyline( mPoints );
+
+    for ( const QgsPoint &point : mPoints )
+    {
+      existingVertices->addGeometry( point.clone() );
+    }
+
+    double perimeter = mDistanceArea.measureLength( geometry );
+    setPerimeter( perimeter );
+    setCanUndo( true );
+  }
+  else
+  {
+    setCanUndo( false );
+  }
+
+  emit existingVerticesChanged( mExistingVertices );
+  setRecordedGeometry( geometry );
+}
+
+void MeasurementMapTool::onMapSettingsChanged( InputMapSettings *newMapSettings )
+{
+  if ( newMapSettings )
```

Comment:
You might need to `disconnect` from the previous mapSettings. You can do so here (in case you override the base function) or disconnect in the base `AbstractMapTool` - I _think_ that should work